internal/http/services/owncloud/ocdav/tus.go

```diff
@@ -0,0 +1,178 @@
+
+package ocdav
+
+import (
+	"net/http"
+	"path"
+
+	rpc "github.com/cs3org/go-cs3apis/cs3/rpc/v1beta1"
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	typespb "github.com/cs3org/go-cs3apis/cs3/types/v1beta1"
+	"github.com/cs3org/reva/pkg/appctx"
+	"github.com/cs3org/reva/pkg/rhttp"
+	tusd "github.com/tus/tusd/pkg/handler"
+)
+
+func (s *svc) handleTusPost(w http.ResponseWriter, r *http.Request, ns string) {
+	ctx := r.Context()
+	log := appctx.GetLogger(ctx)
+
+	w.Header().Add("Access-Control-Allow-Headers", "Tus-Resumable, Upload-Length, Upload-Metadata, If-Match")
+	w.Header().Add("Access-Control-Expose-Headers", "Tus-Resumable, Location")
+
+	w.Header().Set("Tus-Resumable", "1.0.0")
+
+	// Test if the version sent by the client is supported
+	// GET methods are not checked since a browser may visit this URL and does
+	// not include this header. This request is not part of the specification.
+	if r.Header.Get("Tus-Resumable") != "1.0.0" {
+		w.WriteHeader(http.StatusPreconditionFailed)
+		return
+	}
+	if r.Header.Get("Upload-Length") == "" {
+		w.WriteHeader(http.StatusPreconditionFailed)
+		return
+	}
+	//TODO check Expect: 100-continue
+
+	// read filename from metadata
+	meta := tusd.ParseMetadataHeader(r.Header.Get("Upload-Metadata"))
+	if meta["filename"] == "" {
+		w.WriteHeader(http.StatusPreconditionFailed)
+		return
+	}
+
+	ns = applyLayout(ctx, ns)
+
+	// append filename to current dir
+	fn := path.Join(ns, r.URL.Path, meta["filename"])
+
+	// check tus headers?
+
+	// check if destination exists or is a file
+	client, err := s.getClient()
+	if err != nil {
+		log.Error().Err(err).Msg("error getting grpc client")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	sReq := &provider.StatRequest{
+		Ref: &provider.Reference{
+			Spec: &provider.Reference_Path{Path: fn},
+		},
+	}
+	sRes, err := client.Stat(ctx, sReq)
+	if err != nil {
+		log.Error().Err(err).Msg("error sending grpc stat request")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	if sRes.Status.Code != rpc.Code_CODE_OK {
+		if sRes.Status.Code != rpc.Code_CODE_NOT_FOUND {
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+
+	info := sRes.Info
+	if info != nil && info.Type != provider.ResourceType_RESOURCE_TYPE_FILE {
+		log.Warn().Msg("resource is not a file")
+		w.WriteHeader(http.StatusConflict)
+		return
+	}
+
+	if info != nil {
+		clientETag := r.Header.Get("If-Match")
+		serverETag := info.Etag
+		if clientETag != "" {
+			if clientETag != serverETag {
+				log.Warn().Str("client-etag", clientETag).Str("server-etag", serverETag).Msg("etags mismatch")
+				w.WriteHeader(http.StatusPreconditionFailed)
+				return
+			}
+		}
+	}
+
+	// initiateUpload
+
+	uReq := &provider.InitiateFileUploadRequest{
+		Ref: &provider.Reference{
+			Spec: &provider.Reference_Path{Path: fn},
+		},
+		Opaque: &typespb.Opaque{
+			Map: map[string]*typespb.OpaqueEntry{
+				"Upload-Length": {
```

Comment:
CS3APIS extension

Reply:
tracked in https://github.com/cs3org/cs3apis/issues/71